server/src/main/java/org/elasticsearch/common/blobstore/OperationPurpose.java

```diff
@@ -0,0 +1,26 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.common.blobstore;
+
+public enum OperationPurpose {
+    SNAPSHOT("Snapshot"),
+    CLUSTER_STATE("ClusterState"),
+    INDICES("Indices"),
+    TRANSLOG("Translog");
```

Comment:
These are the current set of operation purposes. I think we can start small with these ones. But I am also open to suggestions.

Reply:
I wonder about the name here, which hints that a blob could be used for different purposes, with different operations. I'd be inclined to assume that a blob has a single such purpose, unless we have good examples needing a per operation purpose?

I'd suggest `BlobPurpose` instead.

Reply:
I was under the impression that we want to differentiate between operations (e.g. Get, Put) for the same purpose (e.g. ClusterState) since we both read and write to the cluster state (heartbeat?) blob. So my mental model for the relationship between operation and purpose was combinatory. I could be wrong. But do we want to answer questions like "how much does it cost to maintain translog"? If so, it seems to suggest that we do need to further differentiate "translog" for different operations?

Reply:
I agree that it is combinatoric. I see it as an extra stats field like
```
method="PUT" purpose="translog" count=7
```

but I'd expect a GET against the same file to use the same purpose. With the name `OperationPurpose`, it could be expected to see the same file being accessed with different purposes. I get that there could be some cases of list being indeterminate purpose though (and thus here see a separate one). I find that ok.

> So I updated accordingly.

Thanks.

Reply:
I think we may indeed want to track different operation purposes for individual blobs. For instance we may read a regular snapshot blob for a restore, or for a searchable snapshot (maybe distinguishing prewarming from cache misses) and maybe in future for integrity verification too.

Reply:
Ah, those purposes make sense to me too. I am good with the original name then.

Reply:
David's example is much better and makes a lot sense to me. This confirms that we want to have the combinatory relationship between operation and purpose. It should guide us for the actual structure changes to stats collection which is left out for this PR and will be a follow-up immediately after this one.

For the class name, I kept it to be `BlobPurpose` since I feel it is more generic and can cover different intentions of purpose (be it operational or nominal).

Reply:
Should this be a class/record instead? We only use SNAPSHOT at the moment and the rest is used in plugins. Should it be declared there as well?

Reply:
This Enum does have potential to evolve into an class/interface and offer more functionality. See also my comment [here](https://github.com/elastic/elasticsearch/pull/100236#issuecomment-1746165539). I think we can have follow-ups for it once we have more concrete ideas about usages. We originally wanted it for cost estimation, but then find it can be potentially be used for storage class, priority. We will need consilidate these and maybe more needs.